File: coauthor-for-vs-code/src/textConnection.ts
import OpenAI from 'openai';
import * as vscode from 'vscode';

interface ConnectionResult {
  connector: string;
  choice: string;
}

/**
 * Determines the best way to connect two strings in a LaTeX context using GPT-4
 * @param str1 First string
 * @param str2 Second string
 * @param openaiApiKey Optional OpenAI API key
 * @returns Promise containing the best connector and the model's choice
 */
export async function bestConnectionMethod(
  str1: string,
  str2: string,
  openaiApiKey?: string,
): Promise<ConnectionResult> {
  // If API key not provided, try to get it from VS Code settings
  if (!openaiApiKey) {
    openaiApiKey = vscode.workspace
      .getConfiguration('coauthor.apiKeys')
      .get<string>('openai');

    if (!openaiApiKey) {
      throw new Error(
        'OpenAI API key not found in settings (coauthor.apiKeys.openai)',
      );
    }
  }

  // Define the strings A, B, C
  const A = str1 + str2;
  const B = str1 + ' ' + str2;
  const C = str1 + '\n' + str2;

  // Set up the prompt for the GPT model
  const prompt =
    `Given three strings from a LaTeX document:\n` +
    `A: ${A}\n` +
    `B: ${B}\n` +
    `C: ${C}\n` +
    `Which is more english and latex grammatically correct? Output 'A', 'B', or 'C' directly without giving any reason.`;

  try {
    // Initialize OpenAI client
    const client = new OpenAI({
      apiKey: openaiApiKey,
    });

    // Query the model
    const completion = await client.chat.completions.create({
      model: 'gpt-4-turbo',
      temperature: 0,
      n: 10,
      messages: [
        {
          role: 'system',
          content:
            'You are an assistant trained to determine the most grammatically correct string in a LaTeX document context.',
        },
        {
          role: 'user',
          content: prompt,
        },
      ],
    });

    // Extract and process choices
    const choices = completion.choices.map(
      (choice) => choice.message.content?.trim() ?? '',
    );

    // Determine majority vote
    const choiceCounts = new Map<string, number>();
    choices.forEach((choice) => {
      choiceCounts.set(choice, (choiceCounts.get(choice) ?? 0) + 1);
    });

    let majorityChoice = '';
    let maxCount = 0;
    choiceCounts.forEach((count, choice) => {
      if (count > maxCount) {
        maxCount = count;
        majorityChoice = choice;
      }
    });

    // Map choices to connectors
    const caseDict: { [key: string]: string } = {
      A: '',
      B: ' ',
      C: '\n',
    };

    if (majorityChoice in caseDict) {
      return {
        connector: caseDict[majorityChoice],
        choice: majorityChoice,
      };
    } else {
      console.log(
        `Invalid choice: ${majorityChoice}. Defaulting to adding a space.`,
      );
      return {
        connector: ' ',
        choice: 'B',
      };
    }
  } catch (error) {
    console.error('Error in bestConnectionMethod:', error);
    return {
      connector: ' ',
      choice: 'B',
    };
  }
}

// Example usage:
// async function test() {
//   // const result = await bestConnectionMethod("Hello", "world", "your-api-key");
//   const result = await bestConnectionMethod('Hello', 'world');
//   console.log(result);
// }
// test();

File: coauthor/latex/text_connection.py
import os
from openai import OpenAI

from ..logger import logger


def bestConnectionMethod(str1: str, str2: str, openai_api_key: str | None = None) -> tuple[str, str]:
    """
    Use GPT-4 turbo to determine optimal string concatenation method (direct, space, newline) for LaTeX text.
    Example usage
    str1 = "Hello"
    # str2 = "world"
    # print(bestConnectionMethod(str1, str2))
    """

    # Define the strings A, B, C
    A = str1 + str2
    B = str1 + " " + str2
    C = str1 + "\n" + str2

    # Set up the prompt for the GPT model
    prompt = (
        f"Given three strings from a LaTeX document:\n"
        f"A: {A}\n"
        f"B: {B}\n"
        f"C: {C}\n"
        f"Which is more english and latex grammatically correct? Output 'A', 'B', or 'C' directly without giving any reason."
    )

    # Read API key from environment if not provided
    if openai_api_key is None:
        openai_api_key = os.getenv("OPENAI_API_KEY")

    # Initialize OpenAI client
    client = OpenAI(api_key=openai_api_key)

    # Query the model
    completion = client.chat.completions.create(
        model="gpt-4-turbo",
        temperature=0,
        n=10,
        messages=[
            {
                "role": "developer",
                "content": "You are an assistant trained to determine the most grammatically correct string in a LaTeX document context.",
            },
            {"role": "user", "content": prompt},
        ],
    )

    # Extract the choices from the response
    choices = [choice.message.content.strip() for choice in completion.choices]

    # Determine the majority vote
    majority_choice = max(set(choices), key=choices.count)

    case_dict = {"A": "", "B": " ", "C": "\n"}

    # Return the majority choice directly
    if majority_choice in case_dict:
        return case_dict[majority_choice], majority_choice
    else:
        logger.warning(f"Invalid choice: {majority_choice}. Defaulting to adding a space.")
        return " ", "B"
